### PR TITLE
fix: support image_uri being property ref for model

### PR DESCRIPTION
**Pull request by jerrypeng7773**
*Issue #, if available:*

#3103 

*Description of changes:*

support `image_uri` being property ref for model

*Testing done:*

unit

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backword compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/sagemaker/fw_utils.py [hunk 1]
```diff
@@ -395,8 +397,10 @@ def model_code_key_prefix(code_location_key_prefix, model_name, image):
     Returns:
         str: the key prefix to be used in uploading code
     """
-    training_job_name = sagemaker.utils.name_from_image(image)
-    return "/".join(filter(None, [code_location_key_prefix, model_name or training_job_name]))
+    name_from_image = f"/model_code/{int(time.time())}"
```
Comment (qidewenwhen) [inline src/sagemaker/fw_utils.py:400]: Can we make it a None here?
The line 403 does allow it to be  None. And it's counter intuitive as the default path is not actually a image related name

File: tests/unit/sagemaker/workflow/test_model_step.py [hunk 2]
```diff
@@ -839,3 +841,36 @@ def _verify_register_model_container_definition(
...
+    )
+
+    step_create_model = ModelStep(name="mymodelstep", step_args=model.create())
+
+    pipeline = Pipeline(
+        name="MyPipeline",
+        steps=[lambda_step, step_create_model],
+        sagemaker_session=pipeline_session,
+    )
+    assert pipeline.definition() is not None
```
Comment (qidewenwhen) [inline tests/unit/sagemaker/workflow/test_model_step.py:876]: Can we simply assert that the two properties Get appears in the definition to make the test more strong?
```
        image_uri=lambda_step.properties.Outputs["model_image"],
        model_data=lambda_step.properties.Outputs["model_artifact"],
```

**Review by qidewenwhen — CHANGES_REQUESTED** [review]

File: src/sagemaker/fw_utils.py (hunk 1 above)
Comment (jerrypeng7773) [reply src/sagemaker/fw_utils.py:400]: not sure if I follow, `sagemaker.utils.name_from_image(image)` will always return a value, but `code_location_key_prefix` can be None.

Comment (jerrypeng7773) [reply src/sagemaker/fw_utils.py:400]: as long as its not a pipeline variable, the value will always be overrode by the output from `sagemaker.utils.name_from_image(image)`

File: tests/unit/sagemaker/workflow/test_model_step.py (hunk 2 above)
Comment (jerrypeng7773) [reply tests/unit/sagemaker/workflow/test_model_step.py:876]: sure

File: src/sagemaker/fw_utils.py (hunk 1 above)
Comment (qidewenwhen) [reply src/sagemaker/fw_utils.py:400]: I see, makes sense. Can we use the existing helper function `def name_from_base(base, max_length=63, short=False)` here 
like `name_from_base("model_code")` to make the naming aligned with other modules' convention?
The sagemaker.utils.name_from_image() is calling the `name_from_base` as well

**Review by qidewenwhen — APPROVED** [review]

File: src/sagemaker/fw_utils.py (hunk 1 above)
Comment (jerrypeng7773) [reply src/sagemaker/fw_utils.py:400]: we want the s3 prefix, not `model_code-timestamp` here tho.

**Review by navinsoni — APPROVED** [review]

**Pull request merged**